pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,97 @@
+    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
+    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
+    found in https://arxiv.org/abs/2107.12390.
+
+    Args:
+        frequencies (list): the list of eigenvalue frequencies. Eigenvalue frequencies are defined
+            as the unique positive differences obtained from a set of eigenvalues.
+        diff_shifts (list): the list of shift values. If unspecified, equidistant shifts are
+            assumed. Note that if equidistant frequencies are given, any specified shifts will be
+            ignored, and equidistant shifts will be used instead.
+
+    Returns:
+        tuple: a tuple of one nested list describing the gradient recipe for the parameter-shift
+        method.
+
+    Raises:
+        ValueError: if `frequencies` is not a list of unique positive values, or if `diff_shifts`
+            (if specified) is not a list of unique values the same length as `frequencies`.
+
+    **Example**
+
+    >>> frequencies = [1,2]
+    >>> get_shift_rule(frequencies)
+    ([[0.8535533905932737, 1, 0.7853981633974483], [-0.14644660940672624, 1, 2.356194490192345],
+    [-0.8535533905932737, 1, -0.7853981633974483], [0.14644660940672624, 1, -2.356194490192345]],)
+
+    """
+
+    n_freqs = len(frequencies)
+
+    if not (len(set(frequencies)) == n_freqs and all(freq > 0 for freq in frequencies)):
+        raise ValueError(
+            "Expected frequencies to be a list of unique positive values, instead got {}.".format(
+                frequencies
+            )
+        )
+
+    if diff_shifts is None:  # assume equidistant shifts
+        diff_shifts = [(2 * mu - 1) * np.pi / (2 * n_freqs) for mu in range(1, n_freqs + 1)]
+    else:
+        diff_shifts = list(diff_shifts)
+        if len(diff_shifts) != n_freqs:
+            raise ValueError(
+                "Expected number of shifts to equal the number of frequencies ({}), instead got {}.".format(
+                    n_freqs, diff_shifts
+                )
+            )
+        if len(set(diff_shifts)) != n_freqs:
+            raise ValueError("Shift values must be unique, instead got {}".format(diff_shifts))
+
+    frequencies = sorted(frequencies)
+    freq_min = frequencies[0]
+
+    if np.allclose(
+        np.array(frequencies) / freq_min, range(1, len(frequencies) + 1)
+    ):  # equidistant case
+        diff_shifts = [
+            (2 * mu - 1) * np.pi / (2 * n_freqs * freq_min) for mu in range(1, n_freqs + 1)
+        ]
+        diff_coeffs = [
+            freq_min
+            * (-1) ** (mu - 1)
+            / (4 * n_freqs * np.sin(np.pi * (2 * mu - 1) / (4 * n_freqs)) ** 2)
+            for mu in range(1, n_freqs + 1)
+        ]
+
+    else:  # non-equidistant case
+        sin_matr = -4 * np.sin(np.outer(diff_shifts, frequencies))
+        sin_matr_inv = np.linalg.inv(sin_matr)

```

Comment:
Interesting, I'll look into using `np.linalg.solve`. What frequencies are you specifying to `get_shift_rule` to see the blowup in `sin_matr_inv`?

Reply:
One option, which doesn't fix the instability but rather flags it, is to throw an error in `get_shift_rule` if `np.linalg.det(sin_matr)` is near zero

Reply:
Josh, what are you running this code on that runs into this instability?

Reply:
I just had a look through my IPython history, and I created this instability using a very weird, potentially highly-unlikely, set of frequencies:

```python
>>> frequencies = sorted(np.array([6, 2, 1, 67, 3]))
>>> n_freq = len(frequencies)
>>> freq_min = frequencies[0]
>>> mu = np.arange(1, n_freqs + 1)
>>> shifts = (2 * mu - 1) * np.pi / (2 * n_freqs)
>>> shifts
array([0.31415927, 0.9424778 , 1.57079633, 2.19911486, 2.82743339])
>>> sin_matr = -4 * np.sin(np.outer(shifts, frequencies))
>>> np.round(sin_matr, 4)
array([[-1.2361, -2.3511, -3.2361, -3.8042, -3.2361],
       [-3.2361, -3.8042, -1.2361,  2.3511, -1.2361],
       [-4.    , -0.    ,  4.    , -0.    ,  4.    ],
       [-3.2361,  3.8042, -1.2361, -2.3511, -1.2361],
       [-1.2361,  2.3511, -3.2361,  3.8042, -3.2361]])
>>> sin_matr_inv = np.linalg.inv(sin_matr)
>>> np.round(sin_matr_inv, 4)
array([[-1.13300000e-01,  3.90000000e-03, -1.83600000e-01,
        -0.00000000e+00, -1.13300000e-01],
       [ 2.88000000e-02, -1.84800000e-01,  9.04000000e-02,
         0.00000000e+00,  1.47900000e-01],
       [-2.53011215e+13,  2.53011215e+13, -2.53011215e+13,
         2.53011215e+13, -2.53011215e+13],
       [-1.90200000e-01,  1.53900000e-01, -9.51000000e-02,
         3.63000000e-02, -0.00000000e+00],
       [ 2.53011215e+13, -2.53011215e+13,  2.53011215e+13,
        -2.53011215e+13,  2.53011215e+13]])
>>> coeffs = -2 * np.tensordot(frequencies, sin_matr_inv, axes=1)
>>> coeffs
array([-3.23854355e+15,  3.23854355e+15, -3.23854355e+15,  3.23854355e+15,
       -3.23854355e+15])
```

In this case, determinant is close to 0:

```python
>>> np.linalg.det(sin_matr)
2.8281029073792935e-11
```

Perhaps we could raise a warning if `det(sin_matr)` is close to 0, rather than an outright error?

Reply:
I suppose a simple & sufficient check along these lines would be to make sure that

`(freq_i \pm freq_j)/(2R)` is non-integer for all pairs `freq_i`, `freq_j`. 

If we do observe an integer value for a pair, we can 'squeeze' the equidistant shifts by e.g. scaling the denominator. For example, switching `2R=10` for `12` in the denominator of the shifts restores the numerical stability in the `5` frequency example above!

Reply:
Yeah, one thing that you could do is rescale the frequencies so that the largest frequency is `n_freq`, that should remove this case.

In any case I wonder if it makes sense to remove the default from this case, so we can come back to it later? I'm not super sure equidistant shifts are the sensible default just yet...

Reply:
I'm okay with 'normalizing' the frequencies. @licedric maybe I'm missing it, but what makes you think equidistant shifts is not a good choice for the default?

Reply:
There may very well be better choices for the default that we don't know about yet. For instance, we could choose the shifts to be \pm \pi / (freq + 1) for a given frequency. That seems like it should prevent the singularity, and could perhaps give better scaling as well.

Reply:
This makes a lot of sense!